### PR TITLE
Feature/json handler consumer factory

### LINKED ISSUE #134: Expose debug data for JSON responses
**Is your feature request related to a problem? Please describe.**
The fragment events consumer should expose debug data when response is JSON.

**Describe the solution you'd like**
Consumer gets list of fragments with some execution data. It modifies each fragment (only configured fragments by type) by changing body data.

The processing flow:
- fragment/s supplier -> list of fragments in request context
- fragments handler -> evaluates each fragment
  - http-action
  - payload-to-body
  - here we expect that each fragment contains `JSON` in body
  - engine ends fragment processing
  - consumers are notified
 
Each consumer loads body as JSON and adds execution log to body (and saves).
```json
{
  "user" {
    "profile": "admin"
   }
  ...
  "_knotx_fragment": { // configurable
      "fragment": {},
      "status": {},
      "graph": {}
  }
}
```

When there is only one single fragment, then assembler has nothing to do. 

If there are more fragments with `_knotx_fragment` then we need to apply changes to assembler (in a separate task).

Possible assembler scenarios:
```
[
  {
    "user" { }
    "_knotx_fragment": { }
  },
  {
    "user" { }
    "_knotx_fragment": { }
  }
]
```

**Additional context**
- #79 
- #92

### PR DESCRIPTION
**Pull request by oskarjerzyk**
## Description
Fixes #134.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java [hunk 1]
```diff
@@ -15,9 +15,197 @@
...
+            .put(CONDITION_OPTION, new JsonObject().put(HEADER_OPTION, EXPECTED_HEADER)));
+    tested.accept(new ClientRequest()
+            .setHeaders(MultiMap.caseInsensitiveMultiMap().add(EXPECTED_HEADER, "true")),
+        ImmutableList.of(FragmentExecutionLog.newInstance(original)));
+    //then
+    assertEquals(copy, original);
+  }
+
+  @Test
+  @DisplayName("Fragment should be modified when header condition and supported types configured")
```
Comment (marcinus) [inline handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java]:
```suggestion
  @DisplayName("Fragment should be modified when header condition and fragment type match")
```

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java [hunk 2]
```diff
@@ -15,9 +15,197 @@
...
+    tested.accept(new ClientRequest()
+            .setHeaders(MultiMap.caseInsensitiveMultiMap().add(EXPECTED_HEADER, "true")),
+        ImmutableList.of(FragmentExecutionLog.newInstance(original)));
+
+    //then
+    assertNotEquals(copy, original);
+  }
+
+  @Test
+  @DisplayName("Fragment should be modified when param condition and supported types configured")
```
Comment (marcinus) [inline handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java]:
```suggestion
  @DisplayName("Fragment should be modified when param condition and fragment type match")
```

File: handler/consumer/json/src/main/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactory.java [hunk 3]
```diff
@@ -15,20 +15,87 @@
...
   @Override
   public String getName() {
-    return "json";
+    return "fragmentJsonBodyWriter";
   }
 
   @Override
   public FragmentExecutionLogConsumer create(JsonObject config) {
-    return null;
+    return new FragmentExecutionLogConsumer() {
```
[the remark this answers is not in the document]
Comment (oskarjerzyk) [reply handler/consumer/json/src/main/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactory.java:48]: We have already discussed it with @tomaszmichalak and decided to leave it as it is. For now, these two classes will not have a common part.

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java [hunk 4]
```diff
@@ -48,43 +52,31 @@
...
-    FragmentExecutionLogConsumer tested = new JsonFragmentsHandlerConsumerFactory()
-        .create(new JsonObject()
-            .put(CONDITION_OPTION, new JsonObject().put(HEADER_OPTION, EXPECTED_HEADER)));
-    tested.accept(new ClientRequest()
-            .setHeaders(MultiMap.caseInsensitiveMultiMap().add(EXPECTED_HEADER, "true")),
-        ImmutableList.of(FragmentExecutionLog.newInstance(original)));
-
-    //then
-    assertEquals(copy, original);
+  private static Stream<Arguments> provideFragmentConsumerConfiguration() {
```
Comment (marcinus) [inline handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java:55]: This method does not provide consumer's configuration, but various fragments passed to the consumer being tested. Can we change the name of this method?

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java [hunk 5]
```diff
@@ -48,43 +52,31 @@
...
+    );
   }
 
-  @Test
-  @DisplayName("Fragment body should not be modified when supported fragments do not contain fragment type")
-  void expectFragmentBodyNotModifiedWhenSupportedFragmentsDoNotContainFragmentType() {
-    //given
+  @ParameterizedTest
+  @MethodSource("provideFragmentConsumerConfiguration")
+  @DisplayName("Fragment body should not be modified when invalid configuration provided")
```
Comment (marcinus) [inline handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java:66]:
```suggestion
  @DisplayName("Fragment body should not be modified when no supported methods specified in configuration")
```

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java [hunk 6]
```diff
@@ -48,43 +52,31 @@
...
-            .put(CONDITION_OPTION, new JsonObject().put(HEADER_OPTION, EXPECTED_HEADER)));
-    tested.accept(new ClientRequest()
-            .setHeaders(MultiMap.caseInsensitiveMultiMap().add(EXPECTED_HEADER, "true")),
-        ImmutableList.of(FragmentExecutionLog.newInstance(original)));
-
-    //then
-    assertEquals(copy, original);
+  private static Stream<Arguments> provideFragmentConsumerConfiguration() {
+    return Stream.of(
+        Arguments.of(EXPECTED_FRAGMENT_TYPE, FRAGMENT_BODY_JSON),
```
Comment (marcinus) [inline handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java:57]: Please provide unit test validating that Fragments with HTML body are left unchanged when other conditions match.

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java (hunk 4 above)
Comment (oskarjerzyk) [reply handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java]: fixed

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java (hunk 5 above)
Comment (oskarjerzyk) [reply handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java]: fixed

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java (hunk 6 above)
Comment (oskarjerzyk) [reply handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java]: tests added

File: handler/consumer/json/src/main/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactory.java [hunk 7]
```diff
@@ -15,20 +15,93 @@
...
+        }
+      }
+
+      private boolean isSupported(FragmentExecutionLog executionData) {
+        try {
+          new JsonObject(executionData.getFragment().getBody());
+        } catch (DecodeException e) {
+          return false;
+        }
+        return supportedTypes.contains(executionData.getFragment().getType());
```
Comment (tomaszmichalak) [inline handler/consumer/json/src/main/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactory.java:69]: Maybe we should first check the type :)

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java [hunk 8]
```diff
@@ -15,9 +15,200 @@
...
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
 
-// TODO fix #134
 class JsonFragmentsHandlerConsumerFactoryTest {
 
+  private static final String EXPECTED_FRAGMENT_TYPE = "snippet";
```
Comment (tomaszmichalak) [inline handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java:45]:
```suggestion
  private static final String EXPECTED_FRAGMENT_TYPE = "json";
```

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java [hunk 9]
```diff
@@ -15,9 +15,200 @@
...
+  private static final String EXPECTED_HEADER = "x-knotx-debug";
+  private static final String EXPECTED_PARAM = "debug";
+  private static final String PARAM_OPTION = "param";
+  private static final String HTML_TYPE = "html";
+  private static final String FRAGMENT_BODY_JSON = "{\"user\": \"admin\"}";
+  private static final String FRAGMENT_BODY_HTML = "\"<div>body</div>\"";
+  private static final String USER_KEY = "user";
+  private static final String KNOTX_FRAGMENT_KEY = "_knotx_fragment";
```
Comment (tomaszmichalak) [inline handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java:53]: Please use constant from factory.

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java [hunk 10]
```diff
@@ -15,9 +15,200 @@
...
-// TODO fix #134
 class JsonFragmentsHandlerConsumerFactoryTest {
 
+  private static final String EXPECTED_FRAGMENT_TYPE = "snippet";
+  private static final String EXPECTED_HEADER = "x-knotx-debug";
+  private static final String EXPECTED_PARAM = "debug";
+  private static final String PARAM_OPTION = "param";
+  private static final String HTML_TYPE = "html";
+  private static final String FRAGMENT_BODY_JSON = "{\"user\": \"admin\"}";
+  private static final String FRAGMENT_BODY_HTML = "\"<div>body</div>\"";
```
Comment (tomaszmichalak) [inline handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java:51]:
```suggestion
  private static final String FRAGMENT_BODY_OTHER = "\"simple text value\"";
```

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java [hunk 11]
```diff
@@ -15,9 +15,200 @@
...
+import org.junit.jupiter.params.provider.MethodSource;
 
-// TODO fix #134
 class JsonFragmentsHandlerConsumerFactoryTest {
 
+  private static final String EXPECTED_FRAGMENT_TYPE = "snippet";
+  private static final String EXPECTED_HEADER = "x-knotx-debug";
+  private static final String EXPECTED_PARAM = "debug";
+  private static final String PARAM_OPTION = "param";
+  private static final String HTML_TYPE = "html";
```
Comment (tomaszmichalak) [inline handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java:49]:
```suggestion
  private static final String OTHER_TYPE = "other";
```

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java [hunk 12]
```diff
@@ -15,9 +15,200 @@
...
+  private static final String FRAGMENT_BODY_JSON = "{\"user\": \"admin\"}";
+  private static final String FRAGMENT_BODY_HTML = "\"<div>body</div>\"";
+  private static final String USER_KEY = "user";
+  private static final String KNOTX_FRAGMENT_KEY = "_knotx_fragment";
+  private static final String FRAGMENT_TYPES = "fragmentTypes";
+  private static final String UNSUPPORTED = "unsupported";
+
+  private static Stream<Arguments> provideFragmentTypeAndBody() {
```
Comment (tomaszmichalak) [inline handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java:57]:
```suggestion
  private static Stream<Arguments> unfulfilledConditions() {
```

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java [hunk 13]
```diff
@@ -15,9 +15,200 @@
...
+            .put(FRAGMENT_TYPES, supportedTypes));
+    tested.accept(new ClientRequest()
+            .setHeaders(MultiMap.caseInsensitiveMultiMap().add(EXPECTED_HEADER, "true")),
+        ImmutableList.of(FragmentExecutionLog.newInstance(original)));
+
+    assertEquals(copy, original);
+  }
+
+  @Test
+  @DisplayName("Fragment should not be modified when supported fragments do not contain fragment type")
```
Comment (tomaszmichalak) [inline handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java:95]: Is it this case: `Arguments.of(HTML_TYPE, FRAGMENT_BODY_HTML, new JsonArray().add(EXPECTED_FRAGMENT_TYPE))` ?

File: handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java [hunk 14]
```diff
@@ -15,9 +15,200 @@
...
+
+  @Test
+  @DisplayName("Execution log entry should contain proper fragment details")
+  void expectFragmentDetailsInExecutionLogEntry() {
+    //given
+    FragmentEvent original = new FragmentEvent(
+        new Fragment(EXPECTED_FRAGMENT_TYPE, new JsonObject(), FRAGMENT_BODY_JSON));
+    FragmentEvent copy = new FragmentEvent(original.toJson());
+
+    JsonObject expectedLog = new JsonObject()
```
Comment (tomaszmichalak) [inline handler/consumer/json/src/test/java/io/knotx/fragments/handler/consumer/json/JsonFragmentsHandlerConsumerFactoryTest.java:188]: Please use `FragmentExecutionLog.newInstance(original).toJson()` instead.

File: handler/consumer/api/src/main/java/io/knotx/fragments/handler/consumer/api/model/FragmentExecutionLog.java [hunk 15]
```diff
@@ -96,11 +96,11 @@ public FragmentExecutionLog setStatus(Status status) {
    *
    * @return node processing start time
    */
-  public long getStartTime() {
+  public Long getStartTime() {
```
Comment (tomaszmichalak) [inline handler/consumer/api/src/main/java/io/knotx/fragments/handler/consumer/api/model/FragmentExecutionLog.java:99]: Long -> long

File: handler/core/src/test/resources/task/factory/taskWithNodeLabels.conf [hunk 16]
```diff
@@ -0,0 +1,10 @@
+# node & transitions
```
Comment (tomaszmichalak) [inline handler/core/src/test/resources/task/factory/taskWithNodeLabels.conf:1]: I think labels are in the different PR?

Comment (oskarjerzyk) [reply handler/core/src/test/resources/task/factory/taskWithNodeLabels.conf:1]: of course, it shouldn't be here

**Comment by tomaszmichalak** [comment]
Please create/update README :)

**Pull request merged**